File: native/zohodesk_portal_siq/ios/Classes/ZohodeskPortalSiqPlugin.swift
import Flutter
import UIKit
import ZohoDeskPortalSalesIQ
import ZohoDeskPortalConfiguration

public class ZohodeskPortalSiqPlugin: NSObject, FlutterPlugin {
    
    private var eventSink: FlutterEventSink?
    
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "zohodesk_portal_siq", binaryMessenger: registrar.messenger())
      let eventChannel = FlutterEventChannel(name: "zohodesk_portal_siq_event_channel", binaryMessenger: registrar.messenger())
    let instance = ZohodeskPortalSiqPlugin()
      eventChannel.setStreamHandler(instance)
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch ZDPChatAPIs(rawValue: call.method) {
        case .show:
            ZDPortalSalesIQ.show()
            
        case .setSalesIQInitCallback:
            setSalesIQInitCallback()
            
        case .showOfflineMessage:
            guard let isShow = call.arguments as? Bool else { return }
            ZDPortalSalesIQ.showOfflineMessage(isShow)
            
        case .setGuestUserDetails:
            guard let arguments = call.arguments as? [String: Any] else { return }
            setGuestUserDetails(arguments)
            
        case .setChatBrandDetails:
            guard let arguments = call.arguments as? [String: Any] else { return }
            
        case .setChatVisibility:
            guard let arguments = call.arguments as? [String: Any] else { return }
            setChatVisibility(arguments)
            
        case .hideQueueTime:
            guard let arguement = call.arguments else { return }
            hideQueueTime(arguement)
            
        case .showFeedbackAfterSkip:
            guard let arguement = call.arguments else { return }
            showFeedbackAfterSkip(arguement)
            
        case .enableDragToDismiss:
            guard let arguement = call.arguments else { return }
            enableDragToDismiss(arguement)
            
        case .setKnowledgeBaseVisibility:
            guard let arguement = call.arguments else { return }
            setKnowledgeBaseVisibility(arguement)
            
        case .setLoggerEnabled:
            guard let arguement = call.arguments else { return }
            setLoggerEnabled(arguement)
            
        case .setConversationVisibility:
            guard let arguement = call.arguments else { return }
            setConversationVisibility(arguement)
            
        case .setConversationTitle:
            guard let arguement = call.arguments else { return }
            setConversationTitle(arguement)
            
        case .setLauncherVisibility:
            guard let arguement = call.arguments else { return }
            setLauncherVisibility(arguement)
            
        default:
            break
        }
    }
    
    private func setSalesIQInitCallback() {
        ZDPortalSalesIQ.setCallbackOnInit { [weak self] in
            guard let self else { return }
            eventSink?(true)
        } onError: { [weak self] in
            guard let self else { return }
            eventSink?("Initialization Failed")
        }
    }
    
    private func setGuestUserDetails(_ arguments: [String: Any]) {
        let emailID = arguments["email"] as? String ?? ""
        let displayName = arguments["name"] as? String
        let phoneNumer = arguments["phone"] as? String
        
        ZDPortalSalesIQ.setGuestUser(email: emailID, displayName: displayName, phoneNumber: phoneNumer)
    }
    
    private func setChatBrandDetails(_ arguments: [String: Any]) {
        
        guard let appKey = arguments["appKey"] as? String,
              let accessKey = arguments["accessKey"] as? String
        else { return }
        
        ZDPortalConfiguration.setChatBrandDetails(appKey: appKey, accessKey: accessKey)
    }
    
    private func setChatVisibility(_ arguments: [String: Any]) {
        
        guard let componentValue = arguments["component"] as? String,
              let component = {
                  switch componentValue {
                  case "OPERATOR_IMAGE":
                      return ZDPSalesIQComponents.ChatComponent.attenderImageInChat
                  case "RATING":
                      return .rating
                  case "FEEDBACK":
                      return .feedback
                  case "SCREENSHOT":
                      return .screenshotOption
                  case "PRECHAT_FORM":
                      return .preChatForm
                  case "VISITOR_NAME":
                      return .visitorName
                  case "EMAIL_TRANSCRIPT":
                      return .emailTranscript
                  case "FILE_SHARE":
                      return .fileSharing
                  case "MEDIA_CAPTURE":
                      return .mediaCapture
                  case "END":
                      return .end
                  case "END_WHEN_BOT_CONNECTED":
                      return .endWhenBotConnected
                  case "END_WHEN_OPERATOR_CONNECTED":
                      return .endWhenOperatorConnected
                  case "END_WHEN_IN_QUEUE":
                      return .endWhenInQueue
                  case "REOPEN":
                      return .reopen
                  case "QUEUE_POSITION":
                      return .queuePosition
                  default:
                      return nil
                  }
              }(),
                let visibility = arguments["visibility"] as? Bool
        else { return }
        
        ZDPortalSalesIQ.setChatVisibility(component, visible: visibility)

    }

    private func hideQueueTime(_ arguement: Any) {
        guard let isHide = arguement as? Bool else { return }
        ZDPortalSalesIQ.hideQueueTime(isHide)
    }
    
    private func showFeedbackAfterSkip(_ arguement: Any) {
        guard let isShow = arguement as? Bool else { return }
        ZDPortalSalesIQ.showFeedbackAfterSkip(isShow)
    }
    
    private func enableDragToDismiss(_ arguement: Any) {
        guard let isEnable = arguement as? Bool else { return }
        ZDPortalSalesIQ.enableDragToDismiss(isEnable)
    }
    
    private func setKnowledgeBaseVisibility(_ arguement: Any) {
        guard let isShow = arguement as? Bool else { return }
        ZDPortalSalesIQ.setKnowledgeBaseVisibility(isShow)
    }
    
    private func setLoggerEnabled(_ arguement: Any) {
        guard let isEnable = arguement as? Bool else { return }
        ZDPortalSalesIQ.setLoggerEnabled(isEnable)
    }
    
    private func setConversationVisibility(_ arguement: Any) {
        guard let isShow = arguement as? Bool else { return }
        ZDPortalSalesIQ.setConversationVisibility(isShow)
    }
    
    private func setConversationTitle(_ arguement: Any) {
        guard let title = arguement as? String else { return }
        ZDPortalSalesIQ.setConversationTitle(title)
    }
    
    private func setLauncherVisibility(_ arguement: Any) {
        guard let visibilityName = arguement as? String else { return }
        guard let visibilityMode: ZDPSalesIQComponents.VisibilityMode = {
            switch visibilityName {
            case "ALWAYS_VISIBLE":
                return .always
            case "NEVER_VISIBLE":
                return .never
            case "VISIBLE_WHEN_ACTIVE_CHAT":
                return .whenActiveChat
            default:
                return nil
            }
        }() else { return }
            
        
        ZDPortalSalesIQ.showLauncher(visibilityMode)
    }
    
    private enum ZDPChatAPIs: String {
        case show, showOfflineMessage
        case setSalesIQInitCallback, setGuestUserDetails, setChatBrandDetails
        case setChatVisibility, hideQueueTime, showFeedbackAfterSkip, enableDragToDismiss, setKnowledgeBaseVisibility, setLoggerEnabled, setConversationVisibility
        case setConversationTitle, setLauncherVisibility
    }
}

extension ZohodeskPortalSiqPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}

File: native/zohodesk_portal_siq/android/src/main/kotlin/com/zoho/desk/asap/siq/zohodesk_portal_siq/ZohodeskPortalSiqPlugin.kt
package com.zoho.desk.asap.siq.zohodesk_portal_siq

import android.app.Activity
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import com.zoho.desk.asap.siq.ZohoDeskPortalSalesIQ
import com.zoho.desk.asap.siq.Component
import com.zoho.desk.asap.siq.LauncherMode
import com.zoho.desk.asap.siq.SalesIQInitCallback
import com.google.gson.Gson
import com.zoho.desk.asap.siq.ZDPortalChatUser
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.EventChannel

/** ZohodeskPortalSiqPlugin */
class ZohodeskPortalSiqPlugin: FlutterPlugin, MethodChannel.MethodCallHandler, ActivityAware, EventChannel.StreamHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel
  protected var activity : Activity? = null
  private lateinit var eventChannel : EventChannel
  private var eventSink: EventChannel.EventSink? = null

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "zohodesk_portal_siq")
    channel.setMethodCallHandler(this)
    eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "zohodesk_portal_siq_event_channel");
    eventChannel.setStreamHandler(this)
  }

  override fun onCancel(p0: Any?) {
    eventSink = null
  }

  override fun onListen(p0: Any?, sink: EventChannel.EventSink) {
    eventSink = sink
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
    when(call.method) {
      "show" -> showSalesIQ(call, result)
      "setGuestUserDetails" -> setGuestUserDetails(call, result)
      "setSalesIQInitCallback" -> setSalesIQInitCallback(call, result)
      "setChatBrandDetails" -> setChatBrandDetails(call, result)
      "setChatVisibility" -> setChatVisibility(call, result)
      "showOfflineMessage" -> showOfflineMessage(call, result)
      "hideQueueTime" -> hideQueueTime(call, result)
      "showFeedbackAfterSkip" -> showFeedbackAfterSkip(call, result)
      "enableDragToDismiss" -> enableDragToDismiss(call, result)
      "setKnowledgeBaseVisibility" -> setKnowledgeBaseVisibility(call, result)
      "setLoggerEnabled" -> setLoggerEnabled(call, result)
      "setConversationVisibility" -> setConversationVisibility(call, result)
      "setConversationTitle" -> setConversationTitle(call, result)
      "setLauncherVisibility" -> setLauncherVisibility(call, result)
      else -> result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
    eventSink = null
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onDetachedFromActivityForConfigChanges() {
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
  }

  override fun onDetachedFromActivity() {
    activity = null
  }

  private fun showSalesIQ(call: MethodCall, result: MethodChannel.Result) {
    activity?.let{
      ZohoDeskPortalSalesIQ.show(activity)
    }
  }

  private fun setSalesIQInitCallback(call: MethodCall, result: MethodChannel.Result) {
    ZohoDeskPortalSalesIQ.setSalesIQInitCallback(object: SalesIQInitCallback{
      override fun onInitialized() {
        eventSink?.success(true)
      }

      override fun onException(errorCode: Int, errorMessage: String?) {
        eventSink?.success(errorMessage)
      }
    })
  }

  private fun setGuestUserDetails(call: MethodCall, result: MethodChannel.Result) {
    val paramsMap = call.arguments as? HashMap<String, Any>
    val name = paramsMap?.get("name") as? String?
    val email = paramsMap?.get("email") as? String?
    val phone = paramsMap?.get("phone") as? String?
    val userDetails = ZDPortalChatUser()
    userDetails.name = name
    userDetails.phone = phone
    userDetails.email = email
    ZohoDeskPortalSalesIQ.setGuestUserDetails(userDetails)
  }

  private fun setChatBrandDetails(call: MethodCall, result: MethodChannel.Result) {
    val paramsMap = call.arguments as? HashMap<String, Any>
    val appKey = paramsMap?.get("appKey") as? String?
    val accessKey = paramsMap?.get("accessKey") as? String?

    ZohoDeskPortalSalesIQ.setChatBrandDetails(activity, appKey, accessKey)
  }

  private fun setChatVisibility(call: MethodCall, result: MethodChannel.Result) {
    val paramsMap = call.arguments as? HashMap<String, Any>
    val component = paramsMap?.get("component") as? String?
    val isVisible = paramsMap?.get("visibility") as? Boolean?
    if(component != null && isVisible != null){
      ZohoDeskPortalSalesIQ.setChatVisibility(Component.valueOf(component), isVisible)
    }
  }

  private fun showOfflineMessage(call: MethodCall, result: MethodChannel.Result) {
    (call.arguments as? Boolean?)?.let { isShow ->
      ZohoDeskPortalSalesIQ.showOfflineMessage(isShow)
    }
  }

  private fun hideQueueTime(call: MethodCall, result: MethodChannel.Result) {
    (call.arguments as? Boolean?)?.let { isAllow ->
      ZohoDeskPortalSalesIQ.hideQueueTime(isAllow)
      result.success(true)
    }
  }

  private fun showFeedbackAfterSkip(call: MethodCall, result: MethodChannel.Result) {
    (call.arguments as? Boolean?)?.let { isShow ->
      ZohoDeskPortalSalesIQ.showFeedbackAfterSkip(isShow)
    }
  }

  private fun enableDragToDismiss(call: MethodCall, result: MethodChannel.Result) {
    (call.arguments as? Boolean?)?.let { isEnable ->
      ZohoDeskPortalSalesIQ.enableDragToDismiss(isEnable)
    }
  }

  private fun setKnowledgeBaseVisibility(call: MethodCall, result: MethodChannel.Result) {
    (call.arguments as? Boolean?)?.let { visibility ->
      ZohoDeskPortalSalesIQ.setKnowledgeBaseVisibility(visibility)
    }
  }

  private fun setLoggerEnabled(call: MethodCall, result: MethodChannel.Result) {
    (call.arguments as? Boolean?)?.let { isEnable ->
      ZohoDeskPortalSalesIQ.setLoggerEnabled(isEnable)
    }
  }

  private fun setConversationVisibility(call: MethodCall, result: MethodChannel.Result) {
    (call.arguments as? Boolean?)?.let { visibility ->
      ZohoDeskPortalSalesIQ.setConversationVisibility(visibility)
    }
  }

  private fun setConversationTitle(call: MethodCall, result: MethodChannel.Result) {
    (call.arguments as? String?)?.let { title ->
      ZohoDeskPortalSalesIQ.setConversationTitle(title)
    }
  }

  private fun setLauncherVisibility(call: MethodCall, result: MethodChannel.Result) {
    (call.arguments as? String?)?.let { visibility ->
      ZohoDeskPortalSalesIQ.setLauncherVisibility(LauncherMode.valueOf(visibility))
    }
  }
}
